lib/translator/common.h

```diff
@@ -23,6 +28,34 @@ void DebugLog(const char* format, ...);
 
 void SetDebugCallback(void (*callback)(const char*));
 
+// Max consecutive pages required by NVMe PRP list is 512
+void* AllocPages(uint16_t count);
+
+void* DeallocPages(uint16_t count);
+
+void SetPageCallback(void* (*alloc_callback)(uint16_t),
+                     void* (*dealloc_callback)(uint16_t));
+
+class Translation {
+ public:
+  // Translates from SCSI to NVMe. Translated commands available through
+  // GetNvmeCmds()
+  StatusCode Begin(absl::Span<const uint8_t> scsi_cmd,
+                   scsi_defs::LunAddress lun);
+  // Translates from NVMe to SCSI. Writes SCSI response data to buffer.
+  StatusCode Complete(
+      absl::Span<const nvme_defs::GenericQueueEntryCpl> cpl_data,
+      absl::Span<uint8_t> buffer);
+  // Returns a span containing translated NVMe commands.
+  absl::Span<const nvme_defs::GenericQueueEntryCmd> GetNvmeCmds();
+
+ private:
+  StatusCode pipeline_status;
+  absl::Span<const uint8_t> scsi_cmd;
+  uint32_t nvme_cmd_count;
```

Comment:
im not sure why we have this field but in the next line we have the size of nvme_cmds[] set to 3.

Reply:
The number of valid NVMe cmds will not always be 3, 3 is just estimated as the maximum number we will ever need. Keeping this field allows us to keep track of how many valid commands we should return to the consumer when they call GetNvmeCmds